runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_base.c

```diff
@@ -264,3 +264,118 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
     iree_uk_mmt4d_tile_s8s8s32_2x8x1_arm_64,
     iree_uk_mmt4d_tile_s8s8s32_4x8x1_arm_64,
     iree_uk_mmt4d_tile_s8s8s32_8x8x1_arm_64)
+
+// This kernel is an adaptation of the kernel
+// `qd8-f32-qc4w-gemm-1x16-minmax-neon-mlal-lane.c` in
+// https://github.com/google/XNNPACK. We borrow the int4 conversion trick
+// described within the method body.
+static inline void iree_uk_mmt4d_tile_s8s4s32_1x16x2_to_4x16x2_arm_64(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 1 && M0 <= 4 && iree_uk_is_po2_u32(M0));
+  IREE_UK_ASSERT(!(params->K0 % 2));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+  int32x4_t acc[16];
+  // We start with zero accumulators and add the value of *out_ptr later.
+  // This is required for the int4 trick described later.
+  for (int i = 0; i < 4 * M0; ++i) {
+    acc[i] = vdupq_n_s32(0);
+  }
+
+  const int8x16_t vmask = vmovq_n_s8(0xF0);
+
+  for (int k = 0; k < params->K; ++k) {
+    // Handle 1x16x2.
+    int16_t lhs_0 = *lhs_ptr++;
+    int16_t lhs_1 = *lhs_ptr++;
```

Comment:
Thanks for fixing the out-of-bounds reads, a substantial and important departure from the XNNPACK code (I can now see that the XNNPACK code really is reading out of bounds here, https://github.com/google/XNNPACK/blob/6ef9832d025fb385d69c8fec88e0cb6ffa950e7f/src/qd8-f32-qc4w-gemm/gen/qd8-f32-qc4w-gemm-1x16-minmax-neon-mlal-lane.c#L59 ).

There is a new problem though: the `vmlal_n_s16` intrinsic compiles to the same instruction as `vmlal_lane_s16`, only it also has to generate a second `ins` instruction to insert the RHS scalar into one lane of a NEON register. So, relying on `vmlal_n_s16` for all the arithmetic in this kernel is generally going to lead to suboptimal code, unless the LLVM/Arm back-end can fix that up --- something that we'd rather not rely on if we can avoid to.

Please switch the arithmetic back to `vmlal_lane_s16`, but preserve the fix for out-of-bounds reads my making sure that the RHS NEON register argument to `vmlal_lane_s16` has the relevant lanes populated without out-of-bounds reads. You could do so right here with code doing loads conditionally on `M0`: depending on M0 you can either just to a full `vld1_s8` if `M0==4` or initialize with a `vld1_dup_s8` and complement that with a few `vld1_lane_s8`.

Reply:
yes XNNPack uses an out of bounds read for the remainder handler on input, with a padded weight containing zeros, for 8 bit.
For float the OOB values need to be discarded to avoid denormals or nan, but the read/math are safe for 8 bit.
With 2x16C8, the input read is rounded up to KR which is 8, so if there is a remainder of up to 7, 8 bytes will be read.

Reply:
I re-arranged the loop structure to handle the largest tile size first `4x16x2`. This has the benefit of using `int16x8` and `vld1_s8` to load 8 bytes at a time without any checks. `1x16` and `2x16` uses `int16x4` and loads one lane at a time.

Reply:
int16x8 is 16 bytes?  Did you test performance to check this is faster?